File: lua/problem_22.lua
-- wordBreak reconstructs s from words using backtracking with memoisation.
-- Returns any valid split as a table, or nil if impossible.
local function wordBreak(s, words)
    local wordSet = {}
    for _, w in ipairs(words) do wordSet[w] = true end
    local failed = {}

    local function backtrack(start)
        if start > #s then return {} end  -- fully consumed (1-indexed)
        if failed[start] then return nil end
        for finish = start, #s do
            local word = s:sub(start, finish)
            if wordSet[word] then
                local rest = backtrack(finish + 1)
                if rest ~= nil then
                    local result = { word }
                    for _, w in ipairs(rest) do result[#result+1] = w end
                    return result
                end
            end
        end
        failed[start] = true
        return nil
    end

    return backtrack(1)
end

local function tablesEqual(a, b)
    if #a ~= #b then return false end
    for i = 1, #a do if a[i] ~= b[i] then return false end end
    return true
end

local function problem22()
    print("Problem 22:")
    local r1 = wordBreak("thequickbrownfox", {"quick","brown","the","fox"})
    assert(tablesEqual(r1, {"the","quick","brown","fox"}), "assertion failed")
    local r2 = wordBreak("bedbathandbeyond", {"bed","bath","bedbath","and","beyond"})
    assert(r2 ~= nil, "assertion failed")
    assert(wordBreak("xyz", {"a","b"}) == nil, "assertion failed")
    print("All Tests Passed")
end

return { problem22 = problem22 }

File: python/problem_22.py
def word_break(s: str, words: list[str]) -> list[str] | None:
    # word_break reconstructs s from words using backtracking with memoisation.
    # Returns any valid split, or None if impossible.
    word_set = set(words)
    failed: set[int] = set()

    def backtrack(start: int) -> list[str] | None:
        if start == len(s):
            return []  # fully consumed
        if start in failed:
            return None
        for end in range(start + 1, len(s) + 1):
            word = s[start:end]
            if word in word_set:
                rest = backtrack(end)
                if rest is not None:
                    return [word] + rest
        failed.add(start)
        return None

    return backtrack(0)


def main():
    print("Problem 22:")
    r1 = word_break("thequickbrownfox", ["quick", "brown", "the", "fox"])
    assert r1 == ["the", "quick", "brown", "fox"], "assertion failed"
    r2 = word_break("bedbathandbeyond", ["bed", "bath", "bedbath", "and", "beyond"])
    assert r2 is not None, "assertion failed"
    assert word_break("xyz", ["a", "b"]) is None, "assertion failed"
    print("All Tests Passed")
